File: 1.arrays/merge-intervals/mergeIntervals.js
function mergeIntervals(intervals) {
  intervals = intervals.sort((a, b) => a[0] - b[0]);
  const result = [];
  intervals.forEach((cur) => {
    if (result.length === 0) {
      result.push(cur);
    } else {
      let prev = result[result.length - 1];
      if (prev[prev.length - 1] >= cur[0]) {
        prev[prev.length - 1] = Math.max(
          prev[prev.length - 1],
          cur[cur.length - 1]
        );
      } else {
        result.push(cur);
      }
    }
  });

  return result;
}

const intervals = [
  [1, 3],
  [8, 10],
  [2, 6],
  [15, 18],
];
const intervals2 = [
  [1, 4],
  [4, 5],
];
console.log(mergeIntervals(intervals));
console.log(mergeIntervals(intervals2));

File: 1.arrays/merge-intervals/merge_intervals.rb
def merge_intervals(intervals)
  if intervals.empty? || intervals.length <= 1
    return intervals
  end
  intervals = intervals.sort_by{|x| x.first}
  result = []
  intervals.each do |cur|
    if result.length == 0
      result.push(cur)
    else 
      prev = result.last
      if prev.last >= cur.first
        prev[-1] = [prev.last, cur.last].max
      else
        result.push(cur)
      end
    end
  end
  return result
end



intervals = [[1,3],[8,10],[2,6],[15,18]]
intervals2 = [[1,4],[4,5]]
p(merge_intervals(intervals))
p(merge_intervals(intervals2))